node/pkg/processor/reactor/reactor.go

```diff
@@ -0,0 +1,544 @@
+
+	if c.state.currentState != StateInitialized && c.state.currentState != StateUnobserved && c.state.currentState != StateQuorumUnobserved {
+		return
+	}
+
+	// Late observation
+	if c.state.currentState == StateQuorumUnobserved {
+		reactorObservedLate.WithLabelValues(c.group).Inc()
+	}
+
+	c.state.observation = o
+
+	if c.config.Signer != nil {
+		// Generate digest of the unsigned VAA.
+		digest := o.SigningDigest()
+
+		// Sign the digest using our node's guardian key.
+		timeout, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+		localAddr, err := c.config.Signer.Address(timeout)
+		if err != nil {
+			panic(err)
+		}
+		s, err := c.config.Signer.Sign(timeout, digest.Bytes())
+		if err != nil {
+			panic(err)
+		}
+		c.state.localSignature = s
+
+		messagesSignedTotal.WithLabelValues(c.group).Inc()
+
+		// Store in signatures array
+		c.state.signatures[localAddr] = s
+
+		err = c.transmitSignature(ctx)
+		if err != nil {
+			c.logger.Error("failed to transmit signature on observation", zap.Error(err))
+		}
+
+	}
+
+	// Transition to quorum states
+	switch c.state.currentState {
+	case StateInitialized:
+		c.state.firstSeen = c.clock.Now()
+		c.state.lastObservation = c.clock.Now()
+		c.stateTransition(StateObserved)
+	case StateUnobserved:
+		c.state.lastObservation = c.clock.Now()
+		c.stateTransition(StateObserved)
+	case StateQuorumUnobserved:
+		reactorQuorum.WithLabelValues(c.group, "quorum").Inc()
+		c.state.timeQuorum = c.clock.Now()
+		c.stateTransition(StateQuorum)
+		return
+	}
+
+	// If we haven't reached quorum in this event, there is nothing more to do
+	if !c.hasQuorum() {
+		return
+	}
+
+	// We immediately reached quorum
+	switch c.state.currentState {
+	case StateObserved:
+		reactorQuorum.WithLabelValues(c.group, "quorum").Inc()
+		c.state.timeQuorum = c.clock.Now()
+		c.stateTransition(StateQuorum)
+	}
+}
+
+// timeOut triggers the timeout state transition. It must only be called while holding the reactorState.mutex.
+func (c *ConsensusReactor[K]) timeOut() {
+	if c.state.currentState == StateQuorum {
+		reactorFinalized.WithLabelValues(c.group).Inc()
+		c.stateTransition(StateFinalized)
+	} else {
+		reactorTimedOut.WithLabelValues(c.group, string(c.state.currentState)).Inc()
+		c.stateTransition(StateTimedOut)
+	}
+}
+
+// stateTransition updates the state and triggers the hook. It must only be called while holding the reactorState.mutex.
+func (c *ConsensusReactor[K]) stateTransition(to State) {
+	c.logger.Debug("state transition", zap.String("from", string(c.state.currentState)), zap.String("to", string(to)))
+	previousState := c.state.currentState
+	c.state.currentState = to
+
+	if c.stateTransitionChan != nil {
+		transition := &StateTransition[K]{FromState: previousState, ToState: to, Reactor: c}
+		select {
+		case c.stateTransitionChan <- transition:
+		default:
+			c.logger.Warn("state transition channel full; dropped update", zap.Any("transition", transition))
```

Comment:
Not covered by unit tests. 
Please explain why we'd want to drop transitions here.